iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -66,28 +69,73 @@ cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceD
     {
         auto& entry = m_serviceDescriptions.back();
         entry.emplace(serviceDescription);
+        (*entry).*count = 1U;
         return cxx::success<>();
     }
 
     return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
 }
 
-void ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
+cxx::expected<ServiceRegistry::Error>
+ServiceRegistry::addPublisher(const capro::ServiceDescription& serviceDescription) noexcept
+{
+    return add(serviceDescription, &ServiceDescriptionEntry::publisherCount);
+}
+
+cxx::expected<ServiceRegistry::Error>
+ServiceRegistry::addServer(const capro::ServiceDescription& serviceDescription) noexcept
+{
+    return add(serviceDescription, &ServiceDescriptionEntry::serverCount);
+}
+
+void ServiceRegistry::removePublisher(const capro::ServiceDescription& serviceDescription) noexcept
 {
     auto index = findIndex(serviceDescription);
     if (index != NO_INDEX)
     {
         auto& entry = m_serviceDescriptions[index];
 
-        if (entry->count > 1)
+        if (entry)
         {
-            entry->count--;
+            if (entry->publisherCount >= 1U)
+            {
+                if (--entry->publisherCount >= 1U)
```

Comment:
Let's see if we can keep this after our certification tools run through the codebase :)

What do you think of
```cpp
if(--entry->publisherCount == 0U && entry->serverCount == 0)
{
    entry.reset();
    // reuse the slot in the next insertion
    m_freeIndex = index;
}
```

Reply:
It is not equivalent as in theory you can call it with `entry->publisherCount==0`. In this case decrementing would lead to underflow and would mess everything up.

Reply:
I am indifferent but may change it (do not think in machine code it makes any difference). Regarding certification I do not know, code looks reasonable to me (otherwise I would not have it written like this...).

Reply:
I can also have all conditions in one `if` (short-circuiting) but I am not sure that is nice to read and understand. :-)